File: src/session.ts
// MCP session dedup — the caller-side of episodic memory (epic #31, slice 4).
//
// The other consumer of the artifact log isn't kcp-agent's own history; it's the
// calling agent's MCP session. Its context window IS the session state, so the
// server stays stateless and in character (explicit, not ambient): the caller
// declares the units it already holds — `id → sha256` — and `kcp_load` withholds
// the bytes it would otherwise re-serve, returning an "unchanged" stub instead.
// Real token savings for the caller's window.
//
// The one rule that makes this safe: a stub is only emitted on an EXACT sha
// match. If the caller's copy is stale (the unit drifted), the fresh bytes are
// re-served — an "unchanged" claim is a literal assertion that the bytes are
// identical, never a shortcut that hides a change. And because `kcp_load`
// re-plans (and so re-gates) every call, a unit the caller has since lost access
// to simply isn't in the loaded set — dedup can't smuggle it back as a stub.

import type { LoadedUnit } from "./synthesize.js";

/** What the caller already holds unchanged: an id→sha256 map, or an array of {id, sha256}. */
export type KnownUnits = Record<string, string> | Array<{ id: string; sha256: string }>;

export interface UnchangedUnit {
  id: string;
  path: string;
  sha256: string;
  unchanged: true;
  note: string;
}

export type EmittedUnit = LoadedUnit | UnchangedUnit;

export interface DedupResult {
  units: EmittedUnit[];
  /** Units the caller already had at the same sha — their bytes were withheld. */
  deduped: Array<{ id: string; sha256: string }>;
  /** Content characters withheld — the caller's context-window saving. */
  bytesSaved: number;
}

/** Normalize the caller's declared set into an id→sha256 lookup. */
export function knownMap(known?: KnownUnits): Map<string, string> {
  const m = new Map<string, string>();
  if (!known) return m;
  if (Array.isArray(known)) {
    for (const k of known) if (k && typeof k.id === "string") m.set(k.id, String(k.sha256));
  } else {
    for (const [id, sha] of Object.entries(known)) m.set(id, String(sha));
  }
  return m;
}

/** Withhold the bytes of any loaded unit the caller already holds at the same sha; serve the rest. */
export function dedupeLoaded(loaded: LoadedUnit[], known?: KnownUnits): DedupResult {
  const have = knownMap(known);
  const units: EmittedUnit[] = [];
  const deduped: Array<{ id: string; sha256: string }> = [];
  let bytesSaved = 0;
  for (const u of loaded) {
    if (have.get(u.id) === u.sha256) {
      units.push({
        id: u.id,
        path: u.path,
        sha256: u.sha256,
        unchanged: true,
        note: `unchanged since your copy (sha ${u.sha256.slice(0, 12)}…) — not re-served`,
      });
      deduped.push({ id: u.id, sha256: u.sha256 });
      bytesSaved += u.content.length;
    } else {
      units.push(u);
    }
  }
  return { units, deduped, bytesSaved };
}

File: rust/kcp-planner/src/session.rs
//! Session dedup — a port of `dedupeLoaded` from src/session.ts. A caller passes
//! the units it already holds (`known`, as `[{id, sha256}]` or `{id: sha256}`); a
//! loaded unit whose sha still matches comes back as an "unchanged" stub instead
//! of re-served bytes, saving the caller's context window. Any sha drift re-serves.

use std::collections::HashMap;

use serde_json::{json, Value};

use crate::synthesize::LoadedUnit;

/// Build id→sha256 from either shape MCP callers send.
pub fn known_map(known: Option<&Value>) -> HashMap<String, String> {
    let mut m = HashMap::new();
    match known {
        Some(Value::Array(arr)) => {
            for e in arr {
                if let (Some(id), Some(sha)) = (e.get("id").and_then(|v| v.as_str()), e.get("sha256").and_then(|v| v.as_str())) {
                    m.insert(id.to_string(), sha.to_string());
                }
            }
        }
        Some(Value::Object(obj)) => {
            for (k, v) in obj {
                if let Some(sha) = v.as_str() {
                    m.insert(k.clone(), sha.to_string());
                }
            }
        }
        _ => {}
    }
    m
}

pub struct DedupResult {
    pub units: Vec<Value>,
    pub deduped: Vec<Value>,
    pub bytes_saved: usize,
}

/// Withhold bytes the caller already holds unchanged (exact sha match).
pub fn dedupe_loaded(loaded: &[LoadedUnit], known: Option<&Value>) -> DedupResult {
    let have = known_map(known);
    let mut units = Vec::new();
    let mut deduped = Vec::new();
    let mut bytes_saved = 0usize;
    for u in loaded {
        if have.get(&u.id).map(|s| s == &u.sha256).unwrap_or(false) {
            units.push(json!({
                "id": u.id,
                "path": u.path,
                "sha256": u.sha256,
                "unchanged": true,
                "note": format!("unchanged since your copy (sha {}…) — not re-served", &u.sha256[..12.min(u.sha256.len())]),
            }));
            deduped.push(json!({ "id": u.id, "sha256": u.sha256 }));
            bytes_saved += u.content.encode_utf16().count();
        } else {
            units.push(serde_json::to_value(u).unwrap_or(Value::Null));
        }
    }
    DedupResult { units, deduped, bytes_saved }
}
